lib/logstash/outputs/kafka.rb

```diff
@@ -173,11 +177,20 @@ def register
     @producer = create_producer
     @codec.on_event do |event, data|
       begin
-        if @message_key.nil?
-          record = org.apache.kafka.clients.producer.ProducerRecord.new(event.sprintf(@topic_id), data)
-        else
-          record = org.apache.kafka.clients.producer.ProducerRecord.new(event.sprintf(@topic_id), event.sprintf(@message_key), data)
+        partition = nil
+        key = event.sprintf(@message_key) unless @message_key.nil?
+
+        timestamp = nil
+        unless @message_timestamp.nil?
+          if event.include? @message_timestamp
+            ts = event.get(@message_timestamp)
+            timestamp = ts.is_a?(LogStash::Timestamp) ? ts.time.to_i : ts.to_i
```

Comment:
`LogStash::Timestamp#to_i` returns 'seconds since epoch':

```ruby
[6] pry(#<LogStash::Runner>)> LogStash::Event.new.timestamp.to_i
=> 1505207654
```

But the [Kafka RFC](https://cwiki.apache.org/confluence/display/KAFKA/KIP-32+-+Add+timestamps+to+Kafka+message#KIP-32-AddtimestampstoKafkamessage-Wireprotocolchange-addaTimefieldtothemessageformat) indicates in a few places that the unit for the `Long timestamp` is milliseconds, not seconds.

I think you need `(ts.to_f * 1000).to_i` to convert to milliseconds.

Reply:
Note that I'm not calling `to_i` on a `Logstash::Timestamp`, but on the wrapped Joda `DateTime` object and I was under the impression that back then it returned milliseconds, but agreed, not anymore. I'll change that.

Reply:
Yeah, I checked `timestamp.time` and in Logstash 5.x, at least, it is a ruby Time object:

```
[6] pry(#<LogStash::Runner>)> date = LogStash::Filters::Date.new("match" => [ "message", "ISO8601" ])
=> <LogStash::Filters::Date match=>["message", "ISO8601"], id=>"date_e40418e5-e0ba-4b0f-a3a5-a04b62920d0e", enable_metric=>true, periodic_flush=>false, target=>"@timestamp", tag_on_failure=>["_dateparsefailure"]>
[7] pry(#<LogStash::Runner>)> e = LogStash::Event.new
=> 2017-09-13T04:11:30.230Z %{host} %{message}
[8] pry(#<LogStash::Runner>)> e.timestamp.time.to_i
=> 1505275890
[9] pry(#<LogStash::Runner>)> e.timestamp.time.class
=> Time

# Now try applying the date filter:
[11] pry(#<LogStash::Runner>)> e.set("message", "2017-01-01T00:00:00.000Z")
=> "2017-01-01T00:00:00.000Z"
[12] pry(#<LogStash::Runner>)> date.filter(e)
=> 2017-01-01T00:00:00.000Z %{host} 2017-01-01T00:00:00.000Z
[13] pry(#<LogStash::Runner>)> e.timestamp.time.to_i
=> 1483228800
[14] pry(#<LogStash::Runner>)> e.timestamp.time.class
=> Time

^^^ Still a Time object.
```

Internally, a `LogStash::Event` is a wrapper of the internal org.logstash.Event, and both have a Timestamp class. In Jruby, this timestamp type is provided by JrubyTimestampExtLibrary.java which has a [ruby `time` method that returns a ruby Time object](https://github.com/elastic/logstash/blob/master/logstash-core/src/main/java/org/logstash/ext/JrubyTimestampExtLibrary.java#L118).

So, from Ruby, due to our Jruby wrappers, you don't get to see the `org.logstash.Timestamp` object (which is backed by Joda DateTime type), you see a Ruby `Time` type.

Reply:
True, I made the same exploration and came to the same conclusion. 

Still, I'm a bit surprised that one has to do such a transformation in order to get a millisecond value out of a `LogStash::Timestamp`. Even more surprising is the fact that nowhere else in the whole Logstash codebase something like this is needed. And if my previous statement is wrong (which I hope), then why isn't there a utility method somewhere that gets me that value?

Reply:
I don't know anywhere else we need to output milliseconds-since-epoch. I checked a few places I know we use time that isn't ISO8601 formatted:

* [Nagios output uses seconds-since-epoch (not milliseconds)](https://github.com/logstash-plugins/logstash-output-nagios/blob/master/lib/logstash/outputs/nagios.rb#L96)
* [Graphite output uses seconds-since-epoch](https://github.com/logstash-plugins/logstash-output-graphite/blob/master/lib/logstash/outputs/graphite.rb#L170)
* [Influxdb output supports custom time precision (microseconds up to hours)](https://github.com/logstash-plugins/logstash-output-influxdb/blob/master/lib/logstash/outputs/influxdb.rb#L291-L301)

Reply:
Ok, thanks for checking. I was under the impression that getting a milliseconds representation out of a date/timestamp was a common feature in most mainstream languages. 

I actually found one place in the Logstash core lib that does the same transformation and could benefit from having that factored out into the TimeStamp class: https://github.com/elastic/logstash/blob/master/logstash-core/lib/logstash/agent.rb#L182

If not feasible, I'll just apply your suggestion above so we can let it be.